torch_xla/core/xla_model.py

```diff
@@ -633,6 +636,97 @@ def all_gather(value, dim=0, groups=None, output=None, pin_layout=True):
+                          output=None,
+                          pin_layout=True,
+                          bucket_cap_mb=160):
+  """Performs an all-gather operation along a given dimension, with bucketization.
+
+  Args:
+    See all_gather for the args: dim, groups, output, pin_layout
+    input_list: List of input tensors
+    bucket_cap_mb: Number of MegaBytes of the tensor bucket to fill before doing all-gather.
+
+  Returns:
+    A list of tensors each of which has, in the ``dim`` dimension, all the values from the
+    participating replicas.
+  """
+  # sanity checks
+  if pin_layout:
+    raise RuntimeError(
+        "For xm.all_gather_bucketized, pin_layout=True is not yet supported.")
+  if not isinstance(input_list, list) or any(
+      not isinstance(v, torch.Tensor) for v in input_list):
+    raise TypeError(
+        f"`input_list` needs to be a list of Tensors, but given {type(input_list)}."
+    )
+  if output != None:
+    if not isinstance(output, list) or any(
+        not isinstance(v, torch.Tensor) for v in output):
+      raise TypeError(
+          f"`output` needs to be a list of Tensors, but given {type(output)}.")
+    if len(output) != len(input_list):
+      raise ValueError("`output` length doesn't match `input_list` length: "
+                       f"{len(output)} vs {len(input_list)}.")
+
+  def _all_gather_coalesced(_input_list, _output_list=None):
+    return all_gather(
+        value=_input_list,
+        dim=dim,
+        groups=groups,
+        output=_output_list,
+        pin_layout=pin_layout)
+
+  total = 0
+  tensor_bucket = []
+  output_bucket = [] if output else None
+  out_tensors = []
+  bucket_cap = bucket_cap_mb * 1024 * 1024
+  if groups:
+    divisor = len(groups[0]) if type(groups[0]) == list else len(groups)
+  else:
+    divisor = xrt_world_size()
+  bucket_cap = bucket_cap / divisor
+
+  for idx, tensor in enumerate(input_list):
+    tensor_bytes = tensor.numel() * tensor.element_size()
+
+    # Aim for target bucket_cap_mb: flush new tensor with bucket if bucket content
+    # is small (1/2 cap) but don't combine if combined total is over 2x cap
+    total_new = total + tensor_bytes
+    if tensor_bytes > bucket_cap and total < 0.5 * bucket_cap and total_new <= 2 * bucket_cap:
+      tensor_bucket.append(tensor)
+      if output != None:
+        output_bucket.append(output[idx])
+      out_tensors.extend(_all_gather_coalesced(tensor_bucket, output_bucket))
+      total = 0
+      tensor_bucket = []
+      output_bucket = [] if output else None
+    else:
+      # Bucketize till the total spills over
+      if total_new > bucket_cap:
+        if len(tensor_bucket):
+          out_tensors.extend(
+              _all_gather_coalesced(tensor_bucket, output_bucket))
+        total = 0
+        tensor_bucket = []
+        output_bucket = [] if output else None
+      total = total_new
+      tensor_bucket.append(tensor)
+      if output != None:
+        output_bucket.append(output[idx])
```

Comment:
this part can be refactored a bit it seems like you can refactor out
```
      tensor_bucket.append(tensor)
      if output != None:
        output_bucket.append(output[idx])
```
and
```
        total = 0
        tensor_bucket = []
        output_bucket = [] if output else None
```

it was a bit difficult for me to figure out the difference between these two branches when first reading it.. hopefully refactor would help.

Reply:
refactored